drivers/nrf_802154_sl/src/platform/gpiote/nrf_802154_gpiote_crit_sect.c

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+/**
+ * @file
+ *   This file contains system-agnostic implementation of the nRF 802.15.4 GPIOTE
+ *   critical section abstraction.
+ */
+
+#include <platform/gpiote/nrf_802154_gpiote.h>
+
+#include <hal/nrf_gpio.h>
+#include <nrf_802154_sl_utils.h>
+#include <platform/irq/nrf_802154_irq.h>
+
+/* Whether the GPIOTE interrupt was turned on before entering the first critical section. */
+static volatile bool gpiote_irq_enabled;
+/* Counter of how many times GPIOTE interrupt was disabled while entering critical section. */
+static volatile uint32_t gpiote_irq_disabled_cnt;
+
+void nrf_802154_gpiote_critical_section_enter(void)
+{
+	nrf_802154_sl_mcu_critical_state_t mcu_cs;
+	uint32_t cnt;
+
+	nrf_802154_sl_mcu_critical_enter(mcu_cs);
+	cnt = gpiote_irq_disabled_cnt;
+
+	if (cnt == 0U) {
+		gpiote_irq_enabled = nrf_802154_irq_is_enabled(GPIOTE_IRQn);
+		nrf_802154_irq_disable(GPIOTE_IRQn);
```

Comment:
I don't entirely understand why the whole interrupt is disabled while it seems that only the callback that calls `nrf_802154_wifi_coex_gpiote_irqhandler` should be? Why block servicing of other GPIOTE channels? This does not look well.

Reply:
So if we simply disable callback, and event on given GPIOTE happens and then enable the callback back again. Will it trigger? Will the event "wait" for the callback to be restored. We do this to have critical section. We temporary block the interrupt from GPIOTE do sth that should not be interrupted and restore interrupt back. If you know better solution that will guarantee that event on gpiote will not be missed we are open.

Reply:
And can't you just not perform the call to `nrf_802154_wifi_coex_gpiote_irqhandler` while the critical section is entered and call it on exiting the critical section (if the event occurred in between, obviously)?

Reply:
It would introduce a problem with race conditions, because the context of `nrf_802154_wifi_coex_gpiote_irqhandler` would change from GPIOTE IRQ to an OS thread or another IRQ.

Reply:
What you suggest is a very good improvement. We will take it into account. But please note such change would require excessive stability testing. Please also note that version is marked as 2.0.0.-1.**prealpha** not by a mistake and time is late.